src/subscription/subscribe.js

```diff
@@ -122,44 +148,57 @@ function subscribeImpl(
   // the GraphQL specification. The `execute` function provides the
   // "ExecuteSubscriptionEvent" algorithm, as it is nearly identical to the
   // "ExecuteQuery" algorithm, for which `execute` is also used.
-  return mapAsyncIterator(
-    subscription,
-    payload => execute(
-      schema,
-      document,
-      payload,
-      contextValue,
-      variableValues,
-      operationName,
-      fieldResolver
-    )
+  const mapSourceToResponse = payload => execute(
+    schema,
+    document,
+    payload,
+    contextValue,
+    variableValues,
+    operationName,
+    fieldResolver
+  );
+
+  // Resolve the Source Stream, then map every source value to a 
+  // ExecutionResult value as described above.
+  return sourcePromise.then(
+    sourceStream => mapAsyncIterator(
+      sourceStream,
+      mapSourceToResponse,
+      reportGraphQLError
+    ),
+    reportGraphQLError
   );
 }
 
 /**
  * Implements the "CreateSourceEventStream" algorithm described in the
  * GraphQL specification, resolving the subscription source event stream.
  *
- * Returns an AsyncIterable, may through a GraphQLError.
+ * Returns a Promise<AsyncIterable>.
  *
- * A Source Stream represents the sequence of events, each of which is
- * expected to be used to trigger a GraphQL execution for that event.
+ * If the client-provided invalid arguments, the source stream could not be
+ * created, or the resolver did not return an AsyncIterable, this function will
+ * will throw an error, which should be caught and handled by the caller.
+ *
+ * A Source Event Stream represents a sequence of events, each of which triggers
+ * a GraphQL execution for that event.
  *
  * This may be useful when hosting the stateful subscription service in a
  * different process or machine than the stateless GraphQL execution engine,
  * or otherwise separating these two steps. For more on this, see the
  * "Supporting Subscriptions at Scale" information in the GraphQL specification.
  */
-export function createSourceEventStream(
+export async function createSourceEventStream(
```

Comment:
My only mild concern is that by including an async function in non-test code that we may need to add a dependency to package.json for an async function runtime.

As a follow-up task, let's test this to see if we need to do so, ideally we can ensure no new dependencies.